cirq/google/serializable_device.py

```diff
@@ -0,0 +1,220 @@
+
+        return SerializableDevice(
+            qubits=SerializableDevice._qubits_from_ids(proto.valid_qubits),
+            durations=durations,
+            target_sets=target_sets,
+            permutation_gates=permutation_gates,
+        )
+
+    @staticmethod
+    def _qid_from_str(id_str: str) -> 'cirq.Qid':
+        try:
+            return devices.GridQubit.from_proto_id(id_str)
+        except ValueError:
+            return ops.NamedQubit(id_str)
+
+    @staticmethod
+    def _qubits_from_ids(id_list) -> List['cirq.Qid']:
+        """Translates a list of ids in proto format e.g. '4_3'
+        into cirq.GridQubit objects"""
+        return [SerializableDevice._qid_from_str(id) for id in id_list]
+
+    @staticmethod
+    def _create_target_set(ts: device_pb2.TargetSet
+                          ) -> Set[Tuple['cirq.Qid', ...]]:
+        """Transform a TargetSet proto into a set of qubit tuples"""
+        target_set = set()
+        for target in ts.targets:
+            qid_list = SerializableDevice._qubits_from_ids(target.ids)
+            target_set.add(tuple(qid_list))
+            if ts.target_ordering == device_pb2.TargetSet.SYMMETRIC:
+                qid_list.reverse()
+                target_set.add(tuple(qid_list))
+        return target_set
+
+    def _find_operation_type(self, op_key: 'cirq.Operation',
+                             type_dict: Dict[Type['cirq.Gate'], Any]) -> Any:
+        """Finds the type (or a compatible type) of an operation from within
+        a dictionary with keys of Gate type.
+
+        Returns:
+             the value corresponding to that key or None if no type matches
+        """
+        gate_op = cast(ops.GateOperation, op_key)
+        gate_key = gate_op.gate
+        for type_key in type_dict:
+            if isinstance(gate_key, type_key):
+                return type_dict[type_key]
+        return None
+
+    def duration_of(self, operation: 'cirq.Operation') -> Duration:
+        duration = self._find_operation_type(operation, self.durations)
+        if duration is not None:
+            return duration
+        raise ValueError(
+            f'Operation {operation} does not have a duration listed')
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if self._find_operation_type(operation,
+                                     self.is_permutation_gate) is not None:
+
+            # A permutation gate can have any combination of qubits
+            valid_qubits = self._find_operation_type(operation,
+                                                     self.target_sets)
+            if valid_qubits == set():
+                # All qubits are valid
+                return
+            for q in operation.qubits:
+                for qubit_tuple in valid_qubits:
+                    if q not in qubit_tuple:
```

Comment:
IIUC this raises an error if there is _any_ qubit tuple that does not contain `q`. I would think we'd instead want something like:
```python
valid_target_sets = self._find_operation_type(...)

# flatten into one set; each target _should_ have only one qubit, per device.proto
valid_qubits = {q for target in valid_target_sets for q in target}

if not all(q in valid_qubits for q in operation.qubits):
    raise ValueError(...)
```

Reply:
Done.